crm_opportunity_currency/models/crm_lead.py

```diff
@@ -0,0 +1,65 @@
+# Copyright 2018 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
+
+from odoo import api, fields, models
+
+
+class CrmLead(models.Model):
+    _inherit = 'crm.lead'
+
+    customer_currency_id = fields.Many2one(
+        string='Customer currency',
+        comodel_name='res.currency',
+        default=lambda self: self.env.user.company_id.currency_id,
+    )
+    amount_customer_currency = fields.Monetary(
+        string='Customer amount',
+        currency_field='customer_currency_id',
+        store=True,
+    )
+    is_same_currency = fields.Boolean(
+        string='Same currency',
+        compute='_compute_is_same_currency',
+    )
+
+    @api.onchange('customer_currency_id', 'amount_customer_currency')
+    def _onchange_currency(self):
+        self.planned_revenue = self.get_revenue_in_company_currency()
+
+    @api.multi
+    def get_revenue_in_company_currency(self):
+        """Compute the planned revenue in the company currency.
+
+        If the customer currency is different than the company currency,
+        the planned revenue is computed in the company currency.
+        """
+        self.ensure_one()
+        if self.is_same_currency:
+            return self.planned_revenue
+        return self.customer_currency_id._convert(
+            self.amount_customer_currency or 0,
+            self.company_currency,
+            self.env.user.company_id,
+            fields.Datetime.now(),
+        )
+
+    @api.multi
+    def _compute_is_same_currency(self):
+        for lead in self:
+            lead.is_same_currency = (
+                lead.customer_currency_id == lead.company_currency
+            )
+
+    @api.multi
+    def write(self, vals):
+        res = super().write(vals)
+        if (
+            'customer_currency_id' in vals
+            or 'amount_customer_currency' in vals
+        ):
+            # Due to the readonly in the view the compute does not work
+            for lead in self:
+                lead.write({
```

Comment:
Calling a write inside a loop inside a write is usually a bad design.
I'd add `customer_currency_id=None` and `amount_customer_currency=None` to the params of `get_revenue_in_company_currency`. Then use `or` in the method to get it right, eg:

```
customer_currency = self.env['res.currency'].browse(customer_currency_id) or self.customer_currency_id
return customer_currency._convert(
    amount_customer_currency or self.amount_customer_currency or 0,
    (...)
```
Then move the loop before the call to super and update `vals` for each record, e.g.:
``` 
@api.multi
def write(self, vals):
    if ('customer_currency_id' in vals or 'amount_customer_currency' in vals):
        for lead in self:
            planned_revenue = lead.get_revenue_in_company_currency(
                customer_currency_id=vals.get('customer_currency_id')),
                amount_customer_currency=vals.get('amount_customer_currency')
            )
            vals['planned_revenue'] = planned_revenue
            super(CrmLead, lead).write(vals)
        return True
    else:
        return super().write(vals)
```

Reply:
The change on the write has been removed following changes in the view to handle the problem with the readonly field.